File: intellij-plugin/src/main/kotlin/dev/skillbill/intellij/presentation/SkillBillStatusUiState.kt
package dev.skillbill.intellij.presentation

import dev.skillbill.intellij.domain.CurrentPhaseExecution
import dev.skillbill.intellij.domain.CurrentPhaseModel
import dev.skillbill.intellij.domain.GoalPlanningInfo
import dev.skillbill.intellij.domain.PauseReason
import java.time.Duration
import java.time.Instant

/**
 * Immutable presentation state for Skill Bill status surfaces.
 * No process, JSON, filesystem, or status-bar rendering code here.
 *
 * Authoritative start timestamps ([startedAt] / [subtaskStartedAt]) are retained so
 * UI tickers can recompute elapsed without synthesizing from updated_at. Absent
 * timestamps stay absent; never treat missing as zero.
 */
sealed class SkillBillStatusUiState {
    abstract val headline: String
    abstract val detail: String?
    abstract val goalElapsed: Duration?
    abstract val subtaskElapsed: Duration?
    abstract val progressCompleted: Int?
    abstract val progressTotal: Int?
    abstract val accessibilityText: String
    open val issueKey: String? get() = null
    open val workflowId: String? get() = null
    open val stepLabel: String? get() = null
    open val startedAt: Instant? get() = null
    open val subtaskStartedAt: Instant? get() = null
    open val lastUpdated: Instant? get() = null
    open val problemSummary: String? get() = null

    /**
     * Wire `workflow_family`. Carried to the UI because goal-control eligibility is a
     * function of it; renderers must not re-derive the family from anything else.
     */
    open val workflowFamily: String? get() = null

    /**
     * Snapshot-reported "pause requested, not yet consumed". Null means the snapshot
     * said nothing — never coerced to false, so an absent signal stays distinguishable
     * from an explicit "no pause pending".
     */
    open val pauseRequested: Boolean? get() = null

    /**
     * Non-null only while planning progress is worth showing. Relevance is decided once,
     * in [StatusUiMapper]; renderers must not re-derive it.
     */
    open val planning: GoalPlanningInfo? get() = null

    /** Model the current phase launched with; null on states that carry no current step. */
    open val currentModel: CurrentPhaseModel? get() = null

    open val pauseReason: PauseReason? get() = null

    /**
     * Authoritative current-phase execution measure. Non-null only on lifecycles that carry a
     * current phase; relevance of planning versus this value is decided in [StatusUiMapper].
     */
    open val currentPhaseExecution: CurrentPhaseExecution? get() = null

    /** True when this state is not live. Always true for [Stale]; a modifier elsewhere. */
    open val stale: Boolean get() = this is Stale

    data class Idle(
        override val headline: String = "Skill Bill: idle",
        override val detail: String? = null,
        override val goalElapsed: Duration? = null,
        override val subtaskElapsed: Duration? = null,
        override val progressCompleted: Int? = null,
        override val progressTotal: Int? = null,
        override val lastUpdated: Instant? = null,
        override val stale: Boolean = false,
    ) : SkillBillStatusUiState() {
        override val accessibilityText: String = headline
    }

    data class Done(
        override val headline: String,
        override val detail: String? = null,
        override val goalElapsed: Duration? = null,
        override val subtaskElapsed: Duration? = null,
        override val progressCompleted: Int? = null,
        override val progressTotal: Int? = null,
        override val issueKey: String? = null,
        override val startedAt: Instant? = null,
        override val lastUpdated: Instant? = null,
        override val stale: Boolean = false,
    ) : SkillBillStatusUiState() {
        override val accessibilityText: String = "$headline (done)"
    }

    data class Active(
        override val headline: String,
        override val detail: String?,
        override val goalElapsed: Duration?,
        override val subtaskElapsed: Duration?,
        override val progressCompleted: Int?,
        override val progressTotal: Int?,
        override val issueKey: String?,
        override val workflowId: String?,
        override val stepLabel: String,
        override val startedAt: Instant?,
        override val subtaskStartedAt: Instant?,
        override val lastUpdated: Instant?,
        override val planning: GoalPlanningInfo? = null,
        override val workflowFamily: String? = null,
        override val pauseRequested: Boolean? = null,
        override val currentModel: CurrentPhaseModel? = null,
        override val currentPhaseExecution: CurrentPhaseExecution? = null,
        override val problemSummary: String? = null,
        /** Retained so the 1s ticker re-anchors the active clock without a new poll. */
        val activeDurationMs: Long? = null,
        val activeDurationAsOf: Instant? = null,
        val subtaskActiveDurationMs: Long? = null,
        val subtaskActiveDurationAsOf: Instant? = null,
    ) : SkillBillStatusUiState() {
        override val accessibilityText: String =
            buildString {
                append(headline)
                goalElapsed?.let { append(", goal elapsed ").append(formatDuration(it)) }
                subtaskElapsed?.let { append(", subtask elapsed ").append(formatDuration(it)) }
            }
    }

    data class Paused(
        override val headline: String,
        override val detail: String?,
        override val goalElapsed: Duration?,
        override val subtaskElapsed: Duration?,
        override val progressCompleted: Int?,
        override val progressTotal: Int?,
        override val issueKey: String?,
        override val workflowId: String?,
        override val stepLabel: String,
        override val startedAt: Instant?,
        override val subtaskStartedAt: Instant?,
        override val lastUpdated: Instant?,
        override val planning: GoalPlanningInfo? = null,
        override val workflowFamily: String? = null,
        override val pauseRequested: Boolean? = null,
        override val currentModel: CurrentPhaseModel? = null,
        override val currentPhaseExecution: CurrentPhaseExecution? = null,
        override val pauseReason: PauseReason? = null,
        override val problemSummary: String? = null,
    ) : SkillBillStatusUiState() {
        override val accessibilityText: String = "$headline (paused)"
    }

    data class Stale(
        override val headline: String,
        override val detail: String?,
        override val goalElapsed: Duration?,
        override val subtaskElapsed: Duration?,
        override val progressCompleted: Int?,
        override val progressTotal: Int?,
        override val issueKey: String? = null,
        override val workflowId: String? = null,
        override val stepLabel: String? = null,
        override val startedAt: Instant? = null,
        override val subtaskStartedAt: Instant? = null,
        override val lastUpdated: Instant? = null,
        override val problemSummary: String? = "Cached status is stale",
        override val planning: GoalPlanningInfo? = null,
        override val currentModel: CurrentPhaseModel? = null,
        override val currentPhaseExecution: CurrentPhaseExecution? = null,
    ) : SkillBillStatusUiState() {
        override val accessibilityText: String = "$headline (stale)"
    }

    data class Blocked(
        override val headline: String,
        override val detail: String?,
        override val goalElapsed: Duration?,
        override val subtaskElapsed: Duration?,
        override val progressCompleted: Int? = null,
        override val progressTotal: Int? = null,
        override val issueKey: String? = null,
        override val workflowId: String? = null,
        override val stepLabel: String? = null,
        override val startedAt: Instant? = null,
        override val subtaskStartedAt: Instant? = null,
        override val lastUpdated: Instant? = null,
        override val problemSummary: String? = null,
        override val stale: Boolean = false,
        override val currentModel: CurrentPhaseModel? = null,
        override val currentPhaseExecution: CurrentPhaseExecution? = null,
    ) : SkillBillStatusUiState() {
        override val accessibilityText: String = "$headline (blocked)"
    }

    data class Failed(
        override val headline: String,
        override val detail: String?,
        override val goalElapsed: Duration?,
        override val subtaskElapsed: Duration?,
        override val progressCompleted: Int? = null,
        override val progressTotal: Int? = null,
        override val issueKey: String? = null,
        override val workflowId: String? = null,
        override val stepLabel: String? = null,
        override val startedAt: Instant? = null,
        override val subtaskStartedAt: Instant? = null,
        override val lastUpdated: Instant? = null,
        override val problemSummary: String? = null,
        override val stale: Boolean = false,
        override val currentModel: CurrentPhaseModel? = null,
        override val currentPhaseExecution: CurrentPhaseExecution? = null,
    ) : SkillBillStatusUiState() {
        override val accessibilityText: String = "$headline (failed)"
    }

    data class Unavailable(
        override val headline: String,
        override val detail: String?,
        override val goalElapsed: Duration? = null,
        override val subtaskElapsed: Duration? = null,
        override val progressCompleted: Int? = null,
        override val progressTotal: Int? = null,
        val reasonCode: String,
        override val lastUpdated: Instant? = null,
        override val problemSummary: String? = null,
    ) : SkillBillStatusUiState() {
        override val accessibilityText: String = "$headline (unavailable)"
    }

    data class Incompatible(
        override val headline: String,
        override val detail: String?,
        override val goalElapsed: Duration? = null,
        override val subtaskElapsed: Duration? = null,
        override val progressCompleted: Int? = null,
        override val progressTotal: Int? = null,
        val foundContractVersion: String?,
        override val lastUpdated: Instant? = null,
        override val problemSummary: String? = null,
    ) : SkillBillStatusUiState() {
        override val accessibilityText: String = "$headline (incompatible)"
    }
}

internal fun formatDuration(duration: Duration): String {
    val totalSeconds = duration.seconds.coerceAtLeast(0)
    val hours = totalSeconds / 3_600
    val minutes = (totalSeconds % 3_600) / 60
    val seconds = totalSeconds % 60
    return when {
        hours > 0 -> "%dh %02dm".format(hours, minutes)
        minutes > 0 -> "%dm %02ds".format(minutes, seconds)
        else -> "%ds".format(seconds)
    }
}

File: vscode-extension/src/presentation/SkillBillStatusUiState.ts
import { CurrentPhaseExecution, CurrentPhaseModel, GoalPlanningInfo, PauseReason } from "../domain/SkillBillStatusOutcome";

export function formatDurationMs(durationMs: number): string {
  const totalSeconds = Math.max(0, Math.floor(durationMs / 1000));
  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor((totalSeconds % 3600) / 60);
  const seconds = totalSeconds % 60;
  if (hours > 0) {
    return `${hours}h ${minutes.toString().padStart(2, "0")}m`;
  }
  if (minutes > 0) {
    return `${minutes}m ${seconds.toString().padStart(2, "0")}s`;
  }
  return `${seconds}s`;
}

export type SkillBillStatusUiState =
  | {
      kind: "idle";
      headline?: string;
      detail?: string;
      goalElapsedMs?: number;
      subtaskElapsedMs?: number;
      progressCompleted?: number;
      progressTotal?: number;
      lastUpdated?: Date;
      stale?: boolean;
    }
  | {
      kind: "done";
      headline: string;
      detail?: string;
      goalElapsedMs?: number;
      subtaskElapsedMs?: number;
      progressCompleted?: number;
      progressTotal?: number;
      issueKey?: string;
      startedAt?: Date;
      lastUpdated?: Date;
      stale?: boolean;
    }
  | {
      kind: "active";
      headline: string;
      detail?: string;
      goalElapsedMs?: number;
      subtaskElapsedMs?: number;
      progressCompleted?: number;
      progressTotal?: number;
      issueKey?: string;
      workflowId?: string;
      stepLabel: string;
      startedAt?: Date;
      subtaskStartedAt?: Date;
      lastUpdated?: Date;
      planning?: GoalPlanningInfo;
      workflowFamily?: string;
      pauseRequested?: boolean;
      currentModel?: CurrentPhaseModel;
      currentPhaseExecution?: CurrentPhaseExecution;
      problemSummary?: string;
      activeDurationMs?: number;
      activeDurationAsOf?: Date;
      subtaskActiveDurationMs?: number;
      subtaskActiveDurationAsOf?: Date;
      lastAgentActivityAt?: Date;
      lastAgentActivityLabel?: string;
    }
  | {
      kind: "paused";
      headline: string;
      detail?: string;
      goalElapsedMs?: number;
      subtaskElapsedMs?: number;
      progressCompleted?: number;
      progressTotal?: number;
      issueKey?: string;
      workflowId?: string;
      stepLabel: string;
      startedAt?: Date;
      subtaskStartedAt?: Date;
      lastUpdated?: Date;
      planning?: GoalPlanningInfo;
      workflowFamily?: string;
      pauseRequested?: boolean;
      currentModel?: CurrentPhaseModel;
      currentPhaseExecution?: CurrentPhaseExecution;
      pauseReason?: PauseReason;
      problemSummary?: string;
      lastAgentActivityAt?: Date;
      lastAgentActivityLabel?: string;
    }
  | {
      kind: "stale";
      headline: string;
      detail?: string;
      goalElapsedMs?: number;
      subtaskElapsedMs?: number;
      progressCompleted?: number;
      progressTotal?: number;
      issueKey?: string;
      workflowId?: string;
      stepLabel?: string;
      startedAt?: Date;
      subtaskStartedAt?: Date;
      lastUpdated?: Date;
      problemSummary?: string;
      planning?: GoalPlanningInfo;
      currentModel?: CurrentPhaseModel;
      currentPhaseExecution?: CurrentPhaseExecution;
      lastAgentActivityAt?: Date;
      lastAgentActivityLabel?: string;
    }
  | {
      kind: "blocked";
      headline: string;
      detail?: string;
      goalElapsedMs?: number;
      subtaskElapsedMs?: number;
      progressCompleted?: number;
      progressTotal?: number;
      issueKey?: string;
      workflowId?: string;
      stepLabel?: string;
      startedAt?: Date;
      subtaskStartedAt?: Date;
      lastUpdated?: Date;
      problemSummary?: string;
      stale?: boolean;
      currentModel?: CurrentPhaseModel;
      currentPhaseExecution?: CurrentPhaseExecution;
    }
  | {
      kind: "failed";
      headline: string;
      detail?: string;
      goalElapsedMs?: number;
      subtaskElapsedMs?: number;
      progressCompleted?: number;
      progressTotal?: number;
      issueKey?: string;
      workflowId?: string;
      stepLabel?: string;
      startedAt?: Date;
      subtaskStartedAt?: Date;
      lastUpdated?: Date;
      problemSummary?: string;
      stale?: boolean;
      currentModel?: CurrentPhaseModel;
      currentPhaseExecution?: CurrentPhaseExecution;
    }
  | {
      kind: "unavailable";
      headline: string;
      detail?: string;
      goalElapsedMs?: number;
      subtaskElapsedMs?: number;
      progressCompleted?: number;
      progressTotal?: number;
      reasonCode: string;
      lastUpdated?: Date;
      problemSummary?: string;
    }
  | {
      kind: "incompatible";
      headline: string;
      detail?: string;
      goalElapsedMs?: number;
      subtaskElapsedMs?: number;
      progressCompleted?: number;
      progressTotal?: number;
      foundContractVersion?: string;
      lastUpdated?: Date;
      problemSummary?: string;
    };

export function isStaleState(state: SkillBillStatusUiState): boolean {
  if (state.kind === "stale") {
    return true;
  }
  if ("stale" in state) {
    return state.stale === true;
  }
  return false;
}

export function stateIssueKey(state: SkillBillStatusUiState): string | undefined {
  if ("issueKey" in state) {
    return state.issueKey;
  }
  return undefined;
}

export function stateWorkflowId(state: SkillBillStatusUiState): string | undefined {
  if ("workflowId" in state) {
    return state.workflowId;
  }
  return undefined;
}

export function stateStepLabel(state: SkillBillStatusUiState): string | undefined {
  if ("stepLabel" in state) {
    return state.stepLabel;
  }
  return undefined;
}

export function stateDetail(state: SkillBillStatusUiState): string | undefined {
  if ("detail" in state) {
    return state.detail;
  }
  return undefined;
}

export function stateProblemSummary(state: SkillBillStatusUiState): string | undefined {
  if ("problemSummary" in state) {
    return state.problemSummary;
  }
  return undefined;
}

export function statePlanning(state: SkillBillStatusUiState): GoalPlanningInfo | undefined {
  if ("planning" in state) {
    return state.planning;
  }
  return undefined;
}

export function stateCurrentPhaseExecution(state: SkillBillStatusUiState): CurrentPhaseExecution | undefined {
  if ("currentPhaseExecution" in state) {
    return state.currentPhaseExecution;
  }
  return undefined;
}

export function stateCurrentModel(state: SkillBillStatusUiState): CurrentPhaseModel | undefined {
  if ("currentModel" in state) {
    return state.currentModel;
  }
  return undefined;
}

export function stateWorkflowFamily(state: SkillBillStatusUiState): string | undefined {
  if ("workflowFamily" in state) {
    return state.workflowFamily;
  }
  return undefined;
}

export function stateGoalElapsedMs(state: SkillBillStatusUiState): number | undefined {
  if ("goalElapsedMs" in state) {
    return state.goalElapsedMs;
  }
  return undefined;
}

export function stateSubtaskElapsedMs(state: SkillBillStatusUiState): number | undefined {
  if ("subtaskElapsedMs" in state) {
    return state.subtaskElapsedMs;
  }
  return undefined;
}

export function stateProgressCompleted(state: SkillBillStatusUiState): number | undefined {
  if ("progressCompleted" in state) {
    return state.progressCompleted;
  }
  return undefined;
}

export function stateProgressTotal(state: SkillBillStatusUiState): number | undefined {
  if ("progressTotal" in state) {
    return state.progressTotal;
  }
  return undefined;
}

export function stateLastUpdated(state: SkillBillStatusUiState): Date | undefined {
  if ("lastUpdated" in state) {
    return state.lastUpdated;
  }
  return undefined;
}

export function statePauseReason(state: SkillBillStatusUiState): PauseReason | undefined {
  if ("pauseReason" in state) {
    return state.pauseReason;
  }
  return undefined;
}
